File: java_codes/238. Product of Array Except Self.java
//OPTIMAL TIME COMPLEXITY- O(n) , BEATS 100% , space complexity is O(3n) = O(n) but can be reduced to O(1)

class Solution {
    public int[] productExceptSelf(int[] nums) {

       int prefix[] = new int[nums.length+1];
       int postfix[] = new int[nums.length+1];

        prefix[0] = 1;
        int mul = 1;
       for(int i=0;i<nums.length;i++)
       {
           prefix[i+1] = mul*nums[i];  
           mul = mul*nums[i];
       }


         mul = 1;
        postfix[postfix.length-1]=1;
       for(int i=postfix.length-2;i>=0;i--)
       {
           postfix[i] = mul*nums[i];  
           mul *=nums[i];
       }

       int res[] = new int[nums.length];
       for(int i=0;i<res.length;i++)
       {
           res[i] = prefix[i]*postfix[i+1];
       }
       return res;


    }
}

//OPTIMAL SOLUTION - O(n) - time, O(1) - space

class Solution {
    public int[] productExceptSelf(int[] nums) {

       int n = nums.length;
        int[] res = new int[n];
        // Calculate lefts and store in res.
        int left = 1;
        for (int i = 0; i < n; i++) {
            if (i > 0)
                left = left * nums[i - 1];
            res[i] = left;
        }
        // Calculate rights and the product from the end of the array.
        int right = 1;
        for (int i = n - 1; i >= 0; i--) {
            if (i < n - 1)
                right = right * nums[i + 1];
            res[i] *= right;
        }
        return res;


    }
}

File: python_codes/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1 for _ in range(len(nums)+1)]
        postfix = [1 for _ in range(len(nums)+1)]
        result = [0 for _ in range(len(nums))]

        for i in range (0,len(nums)):
            prefix[i+1] = prefix[i]*nums[i]

        #print(prefix)

        for i in range(len(nums)-1,-1,-1):
            postfix[i] = postfix[i+1]*nums[i]
        
        #print(postfix)

        for i in range(0,len(nums)):
            result[i] = prefix[i]*postfix[i+1]
        return result


# same time complexity but space complexity is O(1)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n

        # prefix product in res
        pref = 1
        for i in range(n):
            res[i] = pref
            pref *= nums[i]

        # multiply by postfix product
        post = 1
        for i in range(n - 1, -1, -1):
            res[i] *= post
            post *= nums[i]

        return res


            